### ISSUE TITLE
Add Documentation for customizing component output with tag prop

### ISSUE THREAD

**Issue by eddywashere** [reactstrap/reactstrap#83]
There should be a page that describes customizing the component tag that is rendered. A great example would be a snippet that documents how to use this with a React Router Link.

```jsx
import React from 'react';
import { Link } from 'react-router';
import { Container, Row, Col, Nav, NavItem, NavLink } from 'reactstrap';

const Example = (props) => {
  return (
    <NavItem>
      <NavLink tag={Link} to="/test" activeClassName="active">
        {props.item.name}
      </NavLink>
    </NavItem>
  );
};
```

**Comment by Harrisonl** [comment]
Yes this is exactly what I am after - I don't want to use a normal href - would much prefer to use the Link. How do I go about doing this?

**Comment by TheSharpieOne** [comment]
Add the tag prop like so: `tag={Link}`

**Comment by eddywashere** [comment]
Yes. That's the way to do it.

**Comment by rstudner** [comment]
Is it possible to do the same for NavbarBrand?  I can't seem to find any combination of things to pass tag that works :)

**Comment by IndifferentDisdain** [comment]
So, I have the following code that seems to be working, except the 'About' isn't clickable since the anchor doesn't render w/ a href attribute; thoughts? Manually navigating to `/#/about` works correctly.

Versions:

- React: 15.4.2
- React Router: 3.0.1
- Reactstrap: 4.0.0-alpha.5

```
import React from 'react';
import {Link} from 'react-router';
import {
  Nav,
  Navbar,
  NavbarBrand,
  NavItem,
  NavLink
} from 'reactstrap';

export default function Menu() {

  return (
    <div>
      <Navbar full inverse>
        <NavbarBrand href="/">Home</NavbarBrand>
        <Nav className="pull-xs-right" navbar>
          <NavItem>
            <NavLink tag={Link} to="/about">About</NavLink>
          </NavItem>
        </Nav>
      </Navbar>
    </div>
  );

}
```

**Comment by eddywashere** [comment]
@IndifferentDisdain does replacing `<NavLink tag={Link} to="/about">About</NavLink>` with <Link to="/about">About</Link>` work? I'm trying to verify this isn't a react-router issue. I had trouble with one of the rr v4 alphas but haven't tried 3.0.1.

**Comment by IndifferentDisdain** [comment]
Nope, that doesn't work either, so I guess it's something in react-router (or I'm doing something wrong w/ it). Will investigate and post here what I find in case anyone else runs into this issue. Thanks!

**Comment by chrise86** [comment]
@IndifferentDisdain did you find a solution? Also facing this issue

**Comment by IndifferentDisdain** [comment]
@chrise86 I did not, sorry :(

**Comment by YutHelloWorld** [comment]
it's good!

**Comment by alekseyl1992** [comment]
It seems, that `tag` should be changed to `NavLink` from `react-router-dom` instead of `Link`.
Original code from first post gives me this: `Warning: Unknown prop `activeClassName` on <a> tag.`

Fixed version:
```
import React from 'react';
import { Link, NavLink as RouterNavLink } from 'react-router';
import { Container, Row, Col, Nav, NavItem, NavLink } from 'reactstrap';

const Example = (props) => {
  return (
    <NavItem>
      <NavLink tag={RouterNavLink} to="/test" activeClassName="active">
        {props.item.name}
      </NavLink>
    </NavItem>
  );
};
```

**Comment by Kizmar** [comment]
FWIW: Using the tag in this way is problematic when using the TypeScript definitions. I was getting the following error when I try to use the tag like that: 
`[ts] Property 'to' does not exist on type 'IntrinsicAttributes & NavLinkProps & { children?: ReactNode; }'.`

I had to add `to?: string` props to both the `NavLinkProps` and `NavbarBrandProps` inferfaces in the corresponding `reactstrap/lib/*.d.ts` files in order for this to work.

**Comment by mohouyizme** [comment]
If you want to use ``NavLink`` instead of ``Link``:
``` js
import { NavLink as Link } from 'react-router-dom';
```

**Comment by Annshuk** [comment]
use NavLink from react-router-dom instead from reactbootstrap and it will work flawless
import {NavLink} from 'react-router-dom';

**Comment by Sharanagouda** [comment]
I have updated react bootstrap to
  "bootstrap": "^4.3.1", 
"react-bootstrap": "^1.0.0-beta.5",

web page is loading for every link 
      <Nav.Item>
               <Nav.Link href="/" eventKey={1} >Home</Nav.Link>
       </Nav.Item>
     <Nav.Item>
              <Nav.Link eventKey={2} href="/about" >About</Nav.Link>
   </Nav.Item>

I have tried 
   <Nav.Link eventKey={2} componentClass={Link}  href="/about" >About</Nav.Link> and 
 <Nav.Link eventKey={2} tag={Link}  to="/about" >About</Nav.Link>  also but not worked

**Comment by endquote** [comment]
I'm trying to set the `tag` to Next.js's `Link from "next/link"`and getting warnings like `Warning: Failed prop type: Link: unknown props found: onClick, className`. But it does actually function correctly.

**Comment by endquote** [comment]
Also: This Next.js starter project uses Reactstrap, but doesn't use `NavLink`, probably for similar reasons:

https://nextjs-starter.now.sh

https://github.com/iaincollins/nextjs-starter/blob/090ae01ecaa10132240ae32545aa09763de91a9d/components/layout.js

It's a little old though.

**Comment by pedox** [comment]
> I'm trying to set the `tag` to Next.js's `Link from "next/link"`and getting warnings like `Warning: Failed prop type: Link: unknown props found: onClick, className`. But it does actually function correctly.

it's kinda weird, since on next.js documentation https://nextjs.org/docs#with-link mentioned to works with Link element, you must place anchor inside the children

```jsx
<Link href='/some/link'>
  <a>Labels</a>
</Link>
```

so, if you want to use NavLink or similar you must wrap it like this i guess

```jsx
<Link href='/some/link'>
  <NavLink>Labels</NavLink>
</Link>
```

**Comment by endquote** [comment]
Trying this:

```jsx
<NavItem>
  <Link href="/posts">
    <NavLink>Posts</NavLink>
  </Link>
</NavItem>
```

Produces this:

```jsx
<li class="nav-item">
  <a class="nav-link">Posts</a>
</li>
```

Note that the `<a>` has no `href`.

If I duplicate the href it seems ok:

```jsx
<NavItem>
  <Link href="/posts">
    <NavLink href="/posts">Posts</NavLink>
  </Link>
</NavItem>
```

**Comment by TheSharpieOne** [comment]
@endquote https://nextjs.org/docs#forcing-the-link-to-expose-href-to-its-child

>If child is an `<a>` tag and doesn't have a `href` attribute we specify it so that the repetition is not needed by the user. However, sometimes, you’ll want to pass an `<a>` tag inside of a wrapper and the **`Link` won’t recognize it as a hyperlink, and, consequently, won’t transfer its href to the child**. In cases like that, you should define a boolean `passHref` property to the `Link`, forcing it to expose its href property to the child.
Please note: using a tag other than `a` and failing to pass `passHref` may result in links that appear to navigate correctly, but, when being crawled by search engines, will not be recognized as links (owing to the lack of href attribute). This may result in negative effects on your sites SEO.

**Comment by endquote** [comment]
Oh, good tip. Thanks for that.

Is `<NavbarBrand>` the same?

```jsx
<Link href="/work" as="/">
  <NavbarBrand>reactstrap</NavbarBrand>
</Link>
```

```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
```

**Comment by TheSharpieOne** [comment]
next/link denote: if passing a functional component as a child of `<Link>` you will need to wrap it in `React.forwardRef`

reactstrap doesn't currently use `React.forwardRef`. There is an open issue to address that #1082.

**Comment by endquote** [comment]
I see. I will look forward to that change. I would like to wrap `NavbarBrand` and `CardLink` with `Link`. I'd also like to set `refs` on `CardImage` and `Card`, but can't right now.

I'll just manually use tags like `<div class="card-link">` in the meantime.

Thanks for your help.

**Comment by meenakshi-dhanani** [comment]
> It seems, that `tag` should be changed to `NavLink` from `react-router-dom` instead of `Link`.
> Original code from first post gives me this: `Warning: Unknown prop `activeClassName` on <a> tag.`
> 
> Fixed version:
> 
> ```
> import React from 'react';
> import { Link, NavLink as RouterNavLink } from 'react-router';
> import { Container, Row, Col, Nav, NavItem, NavLink } from 'reactstrap';
> 
> const Example = (props) => {
>   return (
>     <NavItem>
>       <NavLink tag={RouterNavLink} to="/test" activeClassName="active">
>         {props.item.name}
>       </NavLink>
>     </NavItem>
>   );
> };
> ```

This works for me because of to instead of href. I tried the same thing using react-router-dom's Link. I'm thinking now that maybe the href is the issue and causes the hard reload
